lib/charms/traefik_route_k8s/v0/traefik_route.py

```diff
@@ -140,25 +144,42 @@ def __init__(self, charm: CharmBase, relation_name: str = 'traefik-route'):
                 (defaults to "traefik-route").
         """
         super().__init__(charm, relation_name)
+        self._stored.set_default(external_host=None)
+
         self.charm = charm
-        self.framework.observe(self.charm.on[relation_name].relation_changed,
-                               self._on_relation_changed)
+        self._relation_name = relation_name
+
+        if self._stored.external_host != charm.external_host:
+            self._stored.external_host = charm.external_host
+            self._update_requirers_with_external_host()
```

Comment:
Shouldn't we do this inside a `try..except` just in case the charm does not implement `.external_host`?

Reply:
We have discussed this extensively in the past. This should be an argument to the constructor with a default value. We should never expect certain member variables to exist in charms.

Reply:
This is _explicitly_ the side being passed by Traefik. We _know_ it will exist, and this was the agreed-upon change with the maintainer.

Reply:
We don't know it will exist. This is a charm library and is intended to be used with more multiple charms. In the case of this library in particular, I believe we event *want* other charms to use it right? Sorry to rain on the parade here but I feel pretty strongly this should be a constructor option. I also see no reason why it shouldn't be. It would be easy and would not break the api.

Reply:
I already pushed a patch which moves it to a constructor arg. That said, the argument that it's a charm library and intended to be used with multiple charms is both true and meaningless. Not to put too fine a point on it, but one "half" of charm libraries doesn't need to be that agnostic. [Loki, for example, _assumes_ that there will be a `_container` property on `charm`](https://github.com/canonical/loki-k8s-operator/blob/main/lib/charms/loki_k8s/v0/loki_push_api.py#L1100). I am not saying that it would be better not to, but until there are supervisor/governor charms, some "breaching the veil" with reliance on the fact that we can, in fact, believe that a loki-k8s/traefik-k8s/prometheus-k8s will actually have some attribute and that nobody else is using that "half" of the charm is an assumption we have to live with at times.